benchmark/src/util.ts

```diff
@@ -0,0 +1,84 @@
+import { BenchmarkFunction } from "./benchmark_types";
+
+type Result<T> = { success: true; value: T } | { success: false; error: string };
+
+export function toFixed(num: number, decimalPlaces = 0): string {
+    return string.format(`%.${decimalPlaces}f`, num);
+}
+
+export function calculatePercentageChange(oldValue: number, newValue: number): number {
+    return (newValue / oldValue) * 100 - 100;
+}
+
+// @ts-ignore
+export const isWindows = package.config.sub(1, 1) === "\\";
+
+export const json: {
+    decode: (this: void, str: string) => {};
+    encode: (this: void, val: any) => string;
+} = require("json");
+
+export function readFile(path: string): Result<string> {
+    const fileOpenArray = io.open(path, "rb");
+
+    if (fileOpenArray && fileOpenArray[0]) {
+        const fileHandle = fileOpenArray[0];
+        const readAllResult = readAll(fileHandle);
+        fileHandle.close();
+
+        return readAllResult;
+    }
+    return { success: false, error: `Can't open file ${path}` };
+}
+
+export function readAll(file: LuaFile): Result<string> {
+    const content = file.read(_VERSION === "Lua 5.3" ? "a" : ("*a" as any)) as [string | undefined];
+
+    if (content[0]) {
+        return { success: true, value: content[0] };
+    }
+    return { success: false, error: `Can't readAll for file ${file}` };
+}
+
+export function readDir(dir: string): Result<string[]> {
+    const findHandle = isWindows ? io.popen(`dir /A-D /B ${dir}`) : io.popen(`find '${dir}' -maxdepth 1 -type f`);
+    const findReadAllResult = readAll(findHandle);
+
+    if (!findHandle.close()) {
+        return { success: false, error: `readDir popen failed for dir ${dir} see stdout for more information.` };
```

Comment:
What's the point of using non-standard to the platform monads, when then only thing we can do is propagate the error down on the call stack and die in the end?